sifiveu: add new RISC-V target

wigyori opened the pull request:
This PR adds support for the SiFive U-based, Linux-capable RISC-V SoCs.

* Supported SoCs:
SiFive FU540
SiFive FU740
QEMU support is being prepared for those who don't own actual hardware, and for the StarFive JH SoC which is used on the VisionFive board.

* Supported hardware:
HiFive Unleashed (FU540)
 - CPU: SiFive FU540 quad-core RISC-V (U54, RV64IMAFDC or RV64GC)
 - Memory: 8Gb
 - Ethernet: 1x 10/100/1000

HiFive Unmatched (FU740)
 - CPU: SiFive FU740 quad-core RISC-V (U74, RV64IMAFDCB or RV64GCB)
 - Memory: 16Gb
 - Ethernet: 1x 10/100/1000
 - USB: 4x USB 3.2
 - PCIe:  
     - 1x PCIe Gen3 x8
     - 1x M.2 key M (PCIe x4)
     - 1x M.2 Key E (PCIe x1 / USB2.0)

The OpenWrt core packages and most of the external packages are built successfully with this new architecture (a buildbot was set up to test this new arch), and separate PRs are being submitted for packages which don't (i.e. samba4, perl).

Signed-off-by: Zoltan HERPAI <wigyori@uid0.hu>

wigyori committed:
generic: groundwork for RISC-V

Add build infrastructure for RISC-V.

Signed-off-by: Zoltan HERPAI <wigyori@uid0.hu>

wigyori committed:
sifiveu: add new target for SiFive U-based boards

RISC-V is a new CPU architecture aimed to be fully free and open. This
target will add support for it, based on 5.15.

Supports running on:
 - HiFive Unleashed - FU540, first generation
 - HiFive Unmatched - FU740, current latest generation, PCIe

SD-card images are generated, where the partitions are required to have
specific type codes. As it is commonplace nowadays, OpenSBI is used as the
first stage, with U-boot following as the proper bootloader.

Specifications:

HiFive Unleashed:
 - CPU: SiFive FU540 quad-core RISC-V (U54, RV64IMAFDC or RV64GC)
 - Memory: 8Gb
 - Ethernet: 1x 10/100/1000
 - Console: via microUSB

HiFive Unmatched:
 - CPU: SiFive FU740 quad-core RISC-V (U74, RV64IMAFDCB or RV64GCB)
 - Memory: 16Gb
 - Ethernet: 1x 10/100/1000
 - USB: 4x USB 3.2
 - PCIe:  - 1x PCIe Gen3 x8
          - 1x M.2 key M (PCIe x4)
          - 1x M.2 Key E (PCIe x1 / USB2.0)
 - Console: via microUSB

Installation:
Standard SD-card installation via dd-ing the generated image to
an SD-card of at least 256Mb.

Signed-off-by: Zoltan HERPAI <wigyori@uid0.hu>

wigyori committed:
openssl: add linux-riscv64 into the targets list

Add "linux-riscv64-openwrt" into openssl configurations to enable building
on riscv64.

Signed-off-by: Zoltan HERPAI <wigyori@uid0.hu>

wigyori committed:
opensbi: add package for RISC-V

OpenSBI is a form of a first-stage bootloader, which initializes
certain parts of an SoC and then passes on control to the second
stage bootloader i.e. an u-boot image.

We're introducing the package with release v1.2, which provides
SBI v0.3 and the SBI SRST extensions which helps to gracefully
reboot/shutdown various HiFive-U SoCs.

Tested on SiFive Unleashed and Unmatched boards.

Signed-off-by: Zoltan HERPAI <wigyori@uid0.hu>

wigyori committed:
uboot-sifiveu: add bootloader package for SiFive Ux40 boards

Add new package for building bootloader for the SiFive U-series boards. Supported
boards at this stage are the HiFive Unleashed and HiFive Unmatched.

Signed-off-by: Zoltan HERPAI <wigyori@uid0.hu>

wigyori committed:
firmware-utils: ptgen: add SiFive-related GUID types

Add patch until it gets accepted in firmware-utils upstream.

The SiFive RISC-V SoCs use two special partition types in the boot process.
As a first step, the ZSBL (zero-stage bootloader) in the CPU looks for a
partition with a GUID of 5B193300-FC78-40CD-8002-E86C45580B47 to load the
first-stage bootloader - which in OpenWrt's case is an SPL image. The FSBL
(SPL) then looks for a partition with a GUID of
2E54B353-1271-4842-806F-E436D6AF6985 to load the SSBL which is usually an
u-boot.

With ptgen already supporting GPT partition creation, add the required GUID
types and name them accordingly to be invoked with the '-T <GPT partition
type>' parameter.

Signed-off-by: Zoltan HERPAI <wigyori@uid0.hu>

hauke commented:
I added a new target for the starfive chip here:
https://git.openwrt.org/?p=openwrt/staging/hauke.git;a=shortlog;h=refs/heads/kitchensink-202205-2

This is still based on an older version of your code, I will rebase it on this one in one or two weeks and want to add the code to create a sdcard image. 

I was thing about adding the patches to this target, but I do not know if I break this target. Probably we can merge the two targets in about 2 years when the starfive chip is well supported upstream.

hauke commented on target/linux/sifiveu/base-files/etc/inittab:
```diff
@@ -0,0 +1,6 @@
+tts/0::askfirst:/usr/libexec/login.sh
+ttyS0::askfirst:/usr/libexec/login.sh
+ttySIF0::askfirst:/usr/libexec/login.sh
+tty1::askfirst:/usr/libexec/login.sh
```
Do we need all these uarts?

wigyori replied:
This is leftover from testing, will clean up.

hauke commented on target/linux/sifiveu/config-5.15:
```diff
@@ -0,0 +1,375 @@
+CONFIG_PWM=y
+CONFIG_PWM_SIFIVE=y
+CONFIG_PWM_SYSFS=y
+CONFIG_R8169=y
```
Do you have this realtek Ethernet controller on some boards?

wigyori replied:
R8169 is part of the riscv_defconfig, but indeed it's not needed here (yet).

hauke commented on target/linux/sifiveu/config-5.15:
```diff
@@ -0,0 +1,375 @@
+CONFIG_PADATA=y
+CONFIG_PAGE_OFFSET=0xffffffe000000000
+CONFIG_PAGE_REPORTING=y
+CONFIG_PANIC_TIMEOUT=0
```
generic sets this to `CONFIG_PANIC_TIMEOUT=1`

hauke commented on target/linux/sifiveu/config-5.15:
```diff
@@ -0,0 +1,375 @@
+CONFIG_NEED_DMA_MAP_STATE=y
+CONFIG_NET_FAILOVER=y
+CONFIG_NET_FLOW_LIMIT=y
+CONFIG_NET_NS=y
```
These `_NS` options should be set in the generic config.

wigyori replied:
Ack.

hauke commented on target/linux/sifiveu/image/Config.in:
```diff
@@ -0,0 +1,5 @@
+config SIFIVEU_SD_BOOT_PARTSIZE
```
In the future we should make `TARGET_SD_BOOT_PARTSIZE` a common option shared by multiple target.

hauke replied:
I booted the system build by OpenWrt using the vendor bootloader over the network. 
The SoC has some strange/ different DMA behavior which is not supported upstream. I am not sure if this breaks the SiFive SoCs.
I tried to only backport some good looking patches, but then it booted only sometimes.

hauke replied:
Currently I have all vendor patches, We can definitely remove some of them, but we need some dirty kernel patches.

I haven't looked at the U-Boot and SBI.

cyqtmxk replied:
@hauke @wigyori  Hi all，thanks for sharing your source code.  I'm very interested in openwrt. and I have installed the openwrt on a x86 device. So amazing! I use so many useful packages,like Wake on Lan、 ddns-ipv6、Frp server、vpn、ADguard, etc. I have a VisionFive board too and I want to port the openwrt to the  board, but I dont have any experience，I watched hauke's  presentation  on Youtube （Adding new SoC to OpenWrt by Hauke Mehrtens - OpenWrt Summit）、  I downloaded wigyori‘s codes. I want to use the  vendor linux kernel trees as the externel kernel tree to compile, but  It doesn't work......

I‘ve ported openwrt to VisionFive board for about 4 weeks, but I don’t make any progress，so frustrated for me. 
≡(▔﹏▔)≡

I'm so interested in openwrt and I’m eager to learn from you! Can you give me some suggestions？
Please do not hesitate to contact me if I can help. 
caoyaqing163@qq.com
Thanks！

LiXiaobo1995 replied:
Does there have any guide for visionfive build a sdcard openwrt image?

dangowrt commented on target/linux/sifiveu/image/gen_sifiveu_sdcard_img.sh:
```diff
@@ -0,0 +1,32 @@
+    exit 1
+}
+
+SPL_UUID="5B193300-FC78-40CD-8002-E86C45580B47"
```
This variable seems to be unused

dangowrt commented on target/linux/sifiveu/image/gen_sifiveu_sdcard_img.sh:
```diff
@@ -0,0 +1,32 @@
+}
+
+SPL_UUID="5B193300-FC78-40CD-8002-E86C45580B47"
+UBOOT_UUID="2E54B353-1271-4842-806F-E436D6AF6985"
```
same here

dangowrt commented on target/linux/sifiveu/image/gen_sifiveu_sdcard_img.sh:
```diff
@@ -0,0 +1,32 @@
+
+SPL_UUID="5B193300-FC78-40CD-8002-E86C45580B47"
+UBOOT_UUID="2E54B353-1271-4842-806F-E436D6AF6985"
+BOOT_UUID="EF00"
```
and here

dangowrt commented on target/linux/sifiveu/image/Makefile:
```diff
@@ -0,0 +1,63 @@
+
+define Build/riscv-sdcard
+	rm -f $@.boot #$(KDIR_TMP)/$(IMG_PREFIX)-$(PROFILE)-boot.img
+	mkfs.fat $@.boot  -C $(FAT32_BLOCKS)
```
As we are building U-Boot from source, why use a filesystem from the 1980's to store the kernel?
The boot script be embedded in the default environment of the U-Boot build and uImage.FIT can be stored directly in a partition. In this way the bootloader can validate everything needed for a successful boot before launching the kernel (ie. kernel, dtb **and rootfs**) and reliably dual-boot or launch recovery mechanics in case of integrity errors while avoiding the single point of failure of the FAT filesystem. It also allows to use a single type of image (uImage.FIT) equally on NOR/raw MTD, NAND/UBI and eMMC/block devices.
Examples:
https://github.com/openwrt/openwrt/blob/master/target/linux/mediatek/image/mt7622.mk#L98
https://github.com/openwrt/openwrt/blob/master/package/boot/uboot-mediatek/patches/404-add-bananapi_bpi-r64_defconfigs.patch#L164

I understand that in SBC distributions this aspect is usually neglected as the assumption is that you can always just remove the SD card and stick it into your (Windows) computer. While this is nice to keep the entrance barrier low for tinkerers, the requirements for OpenWrt often look a bit different and involve devices out of the physical reach of their operators...

hauke commented on target/linux/sifiveu/config-5.15:
```diff
@@ -0,0 +1,369 @@
+CONFIG_SYSCTL_EXCEPTION_TRACE=y
+CONFIG_SYSFB=y
+# CONFIG_SYSFB_SIMPLEFB is not set
+CONFIG_SYSFS_SYSCALL=y
```
Do you really need `CONFIG_SYSFS_SYSCALL` ?

hauke commented on target/linux/sifiveu/config-5.15:
```diff
@@ -0,0 +1,369 @@
+CONFIG_PWM=y
+CONFIG_PWM_SIFIVE=y
+CONFIG_PWM_SYSFS=y
+CONFIG_RAS=y
```
Do you really need `CONFIG_RAS` support?

hauke commented on target/linux/sifiveu/config-5.15:
```diff
@@ -0,0 +1,369 @@
+CONFIG_PHYLINK=y
+CONFIG_PHYS_ADDR_T_64BIT=y
+# CONFIG_PHYS_RAM_BASE_FIXED is not set
+CONFIG_PKCS7_MESSAGE_PARSER=y
```
Do you need `CONFIG_PKCS7_MESSAGE_PARSER`?

wigyori replied:
Not necessarily to all three, will clean up later today.

coherent-white commented on tools/firmware-utils/patches/001-add-sifiveu-guid-types.patch:
```diff
@@ -0,0 +1,38 @@
++#define GUID_PARTITION_SIFIVE_UBOOT \
++	GUID_INIT( 0x2e54b353, 0x1271, 0x4842, \
++			0x80, 0x6f, 0xe4, 0x36, 0xd6, 0xaf, 0x69, 0x85)
++
```
Rather than adding the GUIDs, it would be more flexible to move the GUIDs to the invocation script (e.g. gen_sifiveu_sdcard_img.sh) and pass the GUIDs as a parameter.  It would also result in fewer changes to ptgen.c to accept a GUID or "cros_kernel" as a '-T' parameter.

stintel commented:
Why the sifiveu target name instead of using riscv as target name and then sifiveu as subtarget name? If we decide to go for sifiveu as target name, a generic subtarget should be added (see 853e4dd3062df7cb5704b15d6af6730e3194b571).

wigyori commented:
> Why the sifiveu target name instead of using riscv as target name and then sifiveu as subtarget name? If we decide to go for sifiveu as target name, a generic subtarget should be added (see [853e4dd](https://github.com/openwrt/openwrt/commit/853e4dd3062df7cb5704b15d6af6730e3194b571)).

We should stick to sifiveu in my opinion for now, there might/will be other RISC-V SoCs / targets coming up with different feature sets like the D1 has. Fair enough for the generic subtarget - I was just about to push a rebase, let me do another spin with this added.

wigyori commented:
> > Why the sifiveu target name instead of using riscv as target name and then sifiveu as subtarget name? If we decide to go for sifiveu as target name, a generic subtarget should be added (see [853e4dd](https://github.com/openwrt/openwrt/commit/853e4dd3062df7cb5704b15d6af6730e3194b571)).
> 
> We should stick to sifiveu in my opinion for now, there might/will be other RISC-V SoCs / targets coming up with different feature sets like the D1 has. Fair enough for the generic subtarget - I was just about to push a rebase, let me do another spin with this added.

Rebased, added 'generic' subtarget, disabled DA9063 for now (MFD on the Unmatched board).
Ping @stintel .

dangowrt commented:
Looks like that opensbi needs updating to be able to build with newer versions of GCC and binutils:
```diff
diff --git a/package/boot/opensbi/Makefile b/package/boot/opensbi/Makefile
index 7e5287bb54..01348e50aa 100644
--- a/package/boot/opensbi/Makefile
+++ b/package/boot/opensbi/Makefile
@@ -6,13 +6,13 @@
 include $(TOPDIR)/rules.mk
 
 PKG_NAME:=opensbi
-PKG_RELEASE:=1.0
+PKG_RELEASE:=1.2
 
 PKG_SOURCE_PROTO:=git
 PKG_SOURCE_URL=https://github.com/riscv/opensbi
-PKG_SOURCE_DATE:=2021-12-24
-PKG_SOURCE_VERSION:=48f91ee9c960f048c4a7d1da4447d31e04931e38
-PKG_MIRROR_HASH:=4911a3b977d6616787abf6bd8a2299c7552633a1caf2c2bbf39f3e57106e5e4d
+PKG_SOURCE_DATE:=2022-12-24
+PKG_SOURCE_VERSION:=6b5188ca14e59ce7bf71afe4e7d3d557c3d31bf8
+PKG_MIRROR_HASH:=edcdd99da6c62975171981c0aa2b73a27091067da11ccd49816b5ad27d000858
 
 PKG_BUILD_DIR=$(BUILD_DIR)/$(PKG_NAME)-$(BUILD_VARIANT)/$(PKG_NAME)-$(PKG_VERSION)
 
```

Otherwise I'm seeing this:
```
 CC        lib/sbi/sbi_tlb.o
/usr/src/openwrt/build_dir/target-riscv64_riscv64_musl/opensbi-generic/opensbi-2021-12-24-48f91ee9/lib/sbi/sbi_tlb.c: Assembler messages:
/usr/src/openwrt/build_dir/target-riscv64_riscv64_musl/opensbi-generic/opensbi-2021-12-24-48f91ee9/lib/sbi/sbi_tlb.c:190: Error: unrecognized opcode `fence.i', extension `zifencei' required
make[4]: *** [Makefile:431: /usr/src/openwrt/build_dir/target-riscv64_riscv64_musl/opensbi-generic/opensbi-2021-12-24-48f91ee9/build/lib/sbi/sbi_tlb.o] Error 1
```

wigyori commented:
> Looks like that opensbi needs updating to be able to build with newer versions of GCC and binutils:

Builds fine with gcc12+binutils2.39 - hard to test with something that's not in the tree yet :)  Thanks for the heads-up, as the opensbi 1.2 upgrade is already prepped (and tested on the boards) in my staging, will push it here as well.

dangowrt commented:
> Builds fine with gcc12+binutils2.39

I've seen the build error quoted above with gcc12+binututils2.39. 
ry starting with a clean build tree, apply this series on top and realize that neither U-Boot nor OpenSBI will build with gcc12+binutils2.39 as-is, both will need updating,

hauke commented on package/boot/opensbi/Makefile:
```diff
@@ -0,0 +1,63 @@
+PKG_SOURCE_PROTO:=git
+PKG_SOURCE_URL=https://github.com/riscv/opensbi
+PKG_SOURCE_DATE:=2022-12-24
+PKG_SOURCE_VERSION:=6b5188ca14e59ce7bf71afe4e7d3d557c3d31bf8
```
There is also a release:
https://github.com/riscv-software-src/opensbi/releases/download/v1.2/opensbi-1.2-rv-bin.tar.xz

When you use the commit it is hard to know that this is actually version 1.2.

This should download the tar.xz file:
````
PKG_NAME:=opensbi
PKG_RELEASE:=1.2
PKG_RELEASE:=1

PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION)-rv-bin.tar.xz
PKG_SOURCE_URL:=https://github.com/riscv-software-src/opensbi/releases/download/v$(PKG_VERSION)
PKG_HASH:=TODO
````

hauke commented on target/linux/sifiveu/config-5.15:
```diff
@@ -0,0 +1,360 @@
+CONFIG_SG_POOL=y
+CONFIG_SIFIVE_L2=y
+CONFIG_SIFIVE_PLIC=y
+CONFIG_SLUB_DEBUG=y
```
I think this is not needed
```suggestion
```

hauke commented on target/linux/sifiveu/config-5.15:
```diff
@@ -0,0 +1,360 @@
+# CONFIG_USB_EHCI_HCD_PLATFORM is not set
+CONFIG_USB_EHCI_PCI=y
+CONFIG_USB_HID=y
+CONFIG_USB_NET_DRIVERS=y
```
If you do not have USB network devices on the board this is not needed 
```suggestion
```

hauke commented on target/linux/sifiveu/config-5.15:
```diff
@@ -0,0 +1,360 @@
+CONFIG_SPI_MEM=y
+CONFIG_SPI_SIFIVE=y
+CONFIG_SRCU=y
+CONFIG_STACKTRACE=y
```
This is not needed
```suggestion
```

hauke commented on target/linux/sifiveu/config-5.15:
```diff
@@ -0,0 +1,360 @@
+CONFIG_MODULE_SECTIONS=y
+CONFIG_MPILIB=y
+CONFIG_MQ_IOSCHED_DEADLINE=y
+CONFIG_MQ_IOSCHED_KYBER=y
```
not needed
```suggestion
```

hauke commented on target/linux/sifiveu/config-5.15:
```diff
@@ -0,0 +1,360 @@
+CONFIG_MODULES_USE_ELF_RELA=y
+CONFIG_MODULE_SECTIONS=y
+CONFIG_MPILIB=y
+CONFIG_MQ_IOSCHED_DEADLINE=y
```
not needed
```suggestion
```

hauke commented on target/linux/sifiveu/config-5.15:
```diff
@@ -0,0 +1,360 @@
+CONFIG_IRQ_FORCED_THREADING=y
+CONFIG_IRQ_WORK=y
+CONFIG_JBD2=y
+CONFIG_KALLSYMS=y
```
Not needed:
```suggestion
```

hauke commented:
The structure looks good to me.
I also prefer to name this target sifive like you did The vision five already needs different patches and the Allwinner D1 is very different.

aparcar commented:
Could someone refresh the kernel patches please?

dangowrt commented:
@aparcar Afaik development continued here https://git.openwrt.org/?p=openwrt/staging/wigyori.git;a=shortlog;h=refs/heads/riscv-upstream-202301c

wigyori commented:
@aparcar knocking on open doors, I'm refreshing it tonight and then will push the target in.

cyqtmxk commented:
@hauke  Hi hauke,
I boot the "openwrt-visionfive-visionfive-initramfs-fit-uImage.itb" through TFTP, however the Architecture displayed on Status page for StarFive VisionFive V1 is shown as  "?" 
![image](https://user-images.githubusercontent.com/10682818/223007644-34b95ae0-5e65-47cb-b6ea-343e25cc41c8.png)

I opened this about the similar issue: https://github.com/openwrt/luci/issues/5047
Shall we correct it？ thanks !

robimarko commented:
Is there a reason that this target wasn't merged so far?

aparcar commented:
@hauke dont you have hardware for this? Can you have a look?

wigyori commented:
I have been out for most of the last month, will make a refresh and merge it finally.

wigyori commented:
There are a few changes in this latest re-spin, notably:
 - sysupgrade support added
 - use compressed image in FIT
 - u-boot bumped to 2022.10 (as swig is available on the buildbots and other u-boot packages also use it)

The last patch to use binutils 2.38 is a bit of a question mark. When I first tested against the latest 5.15.106 kernel, it initially stuck after started by u-boot at "Starting kernel", and bisecting the versions between the last known-good 5.15.98 and .106 resulted in random stucks also. This initially led to the some patches backported to the previous u-boot/2021.10 release, then to the u-boot upgrade, but the issue remained.
Checking the toolchain, binutils has been since defaulted to 2.40 in trunk, while the Freedom U SDK also uses an earlier binutils release. Reverting back to the known-good binutils 2.38 seems to be working, as I didn't see the issue since. The patch I've added is a conservative approach for now.

aparcar commented:
I just rebased the branch to see if it's working with latest musl update. If it does, I'd be in favour to merge it. CC @hauke @wigyori

hauke commented:
I would also like to see this merged, I think there is a more recent version in @wigyori 's staging tree.

wigyori commented:
- removed the binutils 2.38 downgrade, that was a red herring 
 - switched to use Image instead of uImage for booting
 - tested with various 5.15 kernel versions, they boot fine
 - there was a gcc 12.3 upgrade in the meantime, which might have contributed to fix the issue above
 - tested with the new musl release

This seems to be in a good shape to be merged.

mpratt14 commented:
tools build on CI is erroring for another reason, not actually known

#12700 should fix it though

sorry for the inconvenience

wigyori commented:
- image Makefile cleanup
 - switch to use compressed Image to reduce boot time

Any concerns before merging?

mpratt14 commented:
I don't know anything about sifiveu, I just wanted to let you know about why the CI failed earlier

ynezz reviewed (approved):
LGTM, thanks! BTW feel free to backport it to 23.05 as well.

wigyori commented:
Merged, thanks for all the comments and reviews.

guidosarducci commented:
@wigyori I was wondering about virtual testing options for riscv32/64. It looks like your sifiveu images might run under QEMU, based on machine support. Is that true? Are there options for riscv32? Do we need to add something similar to armvirt perhaps?

Motivation for all this is some periodic BPF subsystem testing I do, and riscv support is new...

wigyori commented:
> @wigyori I was wondering about virtual testing options for riscv32/64. It looks like your sifiveu images might run under QEMU, based on machine support. Is that true? Are there options for riscv32? Do we need to add something similar to armvirt perhaps?
> 
> Motivation for all this is some periodic BPF subsystem testing I do, and riscv support is new...

It's in the plan - currently there are two issues. One is that the MMC emulation is horrendously slow (seems like ARM suffers from this as well), the other is that there is an issue with the i2c initialization.

```
[    1.312390] Oops - load access fault [#1]
[    1.312812] Modules linked in:
[    1.313365] CPU: 0 PID: 1 Comm: swapper/0 Not tainted 5.15.113 #0
[    1.313960] Hardware name: SiFive HiFive Unleashed A00 (DT)
[    1.314535] epc : oc_getreg_8+0x10/0x22
[    1.315556]  ra : ocores_i2c_probe+0xf8/0x684
[    1.315995] epc : ffffffff803d69e2 ra : ffffffff803d7098 sp : ffffffd00400bbb0
[    1.316541]  gp : ffffffff80ecf1e8 tp : ffffffe07ff44980 t0 : ffffffd00400ba40
[    1.317085]  t1 : 000000000000ff00 t2 : 203a65726f636273 s0 : ffffffd00400bbc0
[    1.317618]  s1 : ffffffe07fee6828 a0 : ffffffd00404d008 a1 : 0000000000000008
[    1.318291]  a2 : ffffffff80a34050 a3 : 0000000000000000 a4 : ffffffff80eac086
[    1.319354]  a5 : 0000000000000002 a6 : 0000000000000020 a7 : 0000000000000003
[    1.320056]  s2 : 000000000000000d s3 : ffffffe07fee3c00 s4 : 0000000000000000
[    1.320805]  s5 : ffffffe07fee3c10 s6 : ffffffe080052ba0 s7 : ffffffffffffffea
[    1.321560]  s8 : ffffffe1ffdf1a58 s9 : ffffffff806000ac s10: 0000000000000000
[    1.322227]  s11: 0000000000000000 t3 : 0000000000ff0000 t4 : ffffffffffffffff
[    1.322933]  t5 : 000000ff00000000 t6 : ffffffe080052755
[    1.323515] status: 0000000200000120 badaddr: ffffffd00404d008 cause: 0000000000000005
[    1.324420] [<ffffffff803d69e2>] oc_getreg_8+0x10/0x22
[    1.326590] ---[ end trace 6766da436af20fe0 ]---
[    1.327200] Kernel panic - not syncing: Fatal exception
[    1.327693] SMP: stopping secondary CPUs
[    1.329389] Rebooting in 1 seconds..
```

I'll need to spend some time on the latter.

Apart from these issues, it's quite easy to get it going, you just need the u-boot-spl.bin from the build_dir, which is not installed to bin/targets/sifiveu at this moment.

`qemu-system-riscv64 -M sifive_u,msel=11 -smp 5 -m 8G -display none -serial stdio -bios build_dir/target-riscv64_riscv64_musl/u-boot-sifive_unleashed/u-boot-2022.10/spl/u-boot-spl.bin -drive file=bin/targets/sifiveu/generic/openwrt-sifiveu-generic-sifive_unleashed-ext4-sdcard.img,if=sd`

(You'll need to unzip the sdcard image, and also run a qemu-img resize.)

guidosarducci commented:
Thanks for the update and background, and I'll keep watching for progress.
